crates/noirc_frontend/src/hir/type_check/expr.rs

```diff
@@ -881,38 +920,29 @@ impl<'interner> TypeChecker<'interner> {
 
                 let ret = self.interner.next_type_variable();
                 let args = vecmap(args, |(arg, _, _)| arg);
-                let expected = Type::Function(args, Box::new(ret.clone()));
+                let expected = Type::Function(args, Box::new(ret.clone()), Box::new(Type::Unit));
 
                 if let Err(error) = binding.borrow_mut().bind_to(expected, span) {
                     self.errors.push(error);
                 }
                 ret
             }
-            Type::Function(parameters, ret) => {
-                if parameters.len() != args.len() {
-                    self.errors.push(TypeCheckError::ParameterCountMismatch {
-                        expected: parameters.len(),
-                        found: args.len(),
-                        span,
-                    });
-                    return Type::Error;
-                }
-
-                for (param, (arg, arg_id, arg_span)) in parameters.iter().zip(args) {
-                    arg.make_subtype_with_coercions(
-                        param,
-                        arg_id,
-                        self.interner,
-                        &mut self.errors,
-                        || TypeCheckError::TypeMismatch {
-                            expected_typ: param.to_string(),
-                            expr_typ: arg.to_string(),
-                            expr_span: arg_span,
-                        },
-                    );
-                }
-
-                *ret
+            Type::Function(parameters, ret, env) => {
+                self.bind_function_type_impl(
+                    parameters.as_ref(),
+                    ret.as_ref(),
+                    args.as_ref(),
+                    span,
+                    match *env {
+                        Type::Unit => 0,
+                        Type::Tuple(_) => {
+                            1 // closure env
+                        }
+                        _ => unreachable!(
+                            "function env internal type should be either Unit or Tuple"
+                        ),
```

Comment:
It can be a type variable or named generic as well - but we can remove this check anyway when skip_params is removed

Reply:
you're right, removed